### ISSUE TITLE
Quick notes duplication in the Piano Roll

### ISSUE THREAD

**Issue by unfa** [LMMS/lmms#1876]
I frequently find myself doing one tedious task: copying a bunch of notes and moving them so that the head of the newly created group touches the tail of the previously present notes.

I wish a quick Ctrl+D hotkey could do this for me, it'd save me a lot of Piano-Roll time.

What do you think?

Also: same functionality for Song Editor would rock even more (but not without the ability to remove multiple selected clips at once with Delete key).

**Comment by Sti2nd** [comment]
Yup to this. Should it work for selected notes, like when you have selected a few notes they get duplicated and pasted at the end of the last note in the Piano Roll? Or should they then get pasted at the end of the last selected note?

**Comment by musikBear** [comment]
woulden the most common workflow be:
user made one bar of notes
he wants 4 identical bars
he selects **all** ( ctrl+a
hold ctrl + press D
then all notes inside the selection, would be copied to pos *first empty tick* in *first empty bar* to the right

I remember i have made a similar sugestion where ctrl+ u/d arrow-keys would copy resp one octave up / dn -Also a very common work situation

In *song-editor* this would be an amazing new feature -a real time saver,  so a huge :+1: for that too

**Comment by tresf** [comment]
FYI, Ctrl + D is often a deselect shortcut (also Ctrl + Shift + A)

**Comment by unfa** [comment]
@Stiand
> Yup to this. Should it work for selected notes, like when you have selected a few notes they get duplicated and pasted at the end of the last note in the Piano Roll? Or should they then get pasted at the end of the last selected note?

That's a non-obvious choice to make. I'd say pasting a the end of all notes is safer, as it won't ever produce a situation where the user has to undo the operation to avoid stacked, overlapping notes.

How about a hybrid solution?
LMMS searches for continuous free space in the piano roll that can hold the selected notes, and pastes them here, aligning to the left, so they touch older notes. This way the notes will never overlap, and the user can duplicate in the middle of a long notes sequence, if he has room for it.

Potentially Ctrl+Shift+D could paste immediately after current selection (no matter if there's room or not), but first shift latter notes forward to make space for the new notes. Similar to Shift+Insert in Song Editor:

Visual examples:

AAA - selected notes
BBB - not selected notes
CCC - notes created via duplication

Ctrl+D:

AAABBB
AAABBBCCC

Ctrl+Shift+D:

AAABBB
AAACCCBBB


@tresf
> FYI, Ctrl + D is often a deselect shortcut (also Ctrl + Shift + A)

I didn't know that! In Inkscape and Corel Draw Ctrl+D is "duplicate" though ;)

**Comment by Sti2nd** [comment]
> FYI, Ctrl + D is often a deselect shortcut (also Ctrl + Shift + A)

I added Ctrl + Shift + A to the table as deselect.

**Comment by Sti2nd** [comment]
> I'd say pasting a the end of all notes is safer

And easier to think of / more logical. I think we should wait with extending the functionality to after this has been implemented and see how useful it is.

**Comment by unfa** [comment]
I agree.

**Comment by tresf** [comment]
CTRL + SHIFT + A has kindly been implemented by @Wallacoloo via #1963 (#1488)

**Comment by musikBear** [comment]
:+1:
